Readme.md
# retext-link [![Build Status](https://img.shields.io/travis/wooorm/retext-link.svg?style=flat)](https://travis-ci.org/wooorm/retext-link) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-link.svg?style=flat)](https://coveralls.io/r/wooorm/retext-link?branch=master)

**[retext](https://github.com/wooorm/retext "Retext")** link support.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-link
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-link
```

[Duo](http://duojs.org/#getting-started):

```javascript
var link = require('wooorm/retext-link');
```

## Usage

```javascript
var Retext = require('retext');
var link = require('retext-link');
var visit = require('retext-visit');

var retext = new Retext()
    .use(visit)
    .use(link);

retext.parse(
    'Have you seen http://foo.com/blah_blah_(wikipedia)_(again)? ' +
    'It’s pretty cool. Send me an email at mailto:test+foo@bar.com',
    function (err, tree) {
        tree.visit(tree.LINK_NODE, function (node) {
            console.log('Link', node.toString());
            console.log('Relative?', node.isRelative());
            console.log('Info', node.data);
            console.log();
        });
        /*
         * Node 'http://foo.com/blah_blah_(wikipedia)_(again)'
         * Relative? false
         * Info {
         *   protocol: 'http:',
         *   host: 'foo.com',
         *   port: 80,
         *   hostname: 'foo.com',
         *   hash: '',
         *   search: '',
         *   query: '',
         *   pathname: '/blah_blah_(wikipedia)_(again)',
         *   href: 'http://foo.com/blah_blah_(wikipedia)_(again)'
         * }
         *
         * Node 'mailto:test+foo@bar.com'
         * Relative? true
         * Info {
         *   protocol: 'mailto:',
         *   host: 'bar.com',
         *   port: 80,
         *   hostname: 'bar.com',
         *   hash: '',
         *   search: '',
         *   query: '',
         *   pathname: '',
         *   href: 'mailto:test%2Bfoo@bar.com'
         * }
         */
    }
);
```

## API

**retext-link** automatically detects links in a document and classifies them as [`LinkNode`](https://github.com/wooorm/textom-link-node#textomlinknode)s.

## Performance

On a MacBook Air. **retext** works 9% faster on content with links, and 2% slower on content without links, when using **retext-link**.

```text
           retext w/o retext-link
  166 op/s » A paragraph (5 sentences, 100 words, 10 links)
  238 op/s » A paragraph (5 sentences, 100 words, no links)

           retext w/ retext-link
  152 op/s » A paragraph (5 sentences, 100 words, 10 links)
  244 op/s » A paragraph (5 sentences, 100 words, no links)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.2.4 / 2015-01-18
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update retext
  * Update eslint

0.2.3 / 2014-12-14
==================

 * Add benchmark
 * Fix wording in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Update textom-link-node, eslint, retext

0.2.2 / 2014-11-22
==================

 * Refactor `Readme.md`
 * Add flat badges to `Readme.md`
 * Refactor npm script targets in `package.json`
 * Remove retext-visit, retext-cst from dev-dependencies
 * Update retext, retext-inspect, retext-cst

0.2.1 / 2014-11-19
==================

 * Update nlcst-link-modifier

0.2.0 / 2014-11-19
==================

 * Add new LinkNodes, better parser
 * Add retext-inspect as a dev-dependency
 * Add nlcst-link-modifier, textom-link-node as dependencies
 * Remove bower support
 * Fix executive rights on `test.js`
 * Update retext

0.1.3 / 2014-10-28
==================

 * Add `.eslintrc`
 * Add nlcst-to-string as a dependency
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext, retext-ast, retext-visit

0.1.2 / 2014-10-21
==================

 * Update module for changes in retext
 * Refactor .jscs.json
 * Update retext

0.1.1 / 2014-10-15
==================

 * Update tests for changes to retext-visit
 * Update retext, retext-ast, retext-visit
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js

0.1.0 / 2014-09-26
==================

 * Refactor Readme.md
 * Refactor API
 * Add npm deployment to travis
 * Remove `npm update` from travis
 * Remove browser test
 * Update .gitignore, .npmignore
 * Update Installation in docs
 * Add bower.json
 * Fix property order in component.json, package.json
 * Update copyright in Readme.md
 * Remove testling
 * Refactor spec
 * Update dependencies
 * Add retext as a peer dependency
 * Add components to gitignore


test.js
'use strict';

/**
 * Dependencies.
 */

var link,
    inspect,
    Retext,
    assert;

link = require('./');
inspect = require('retext-inspect');
Retext = require('retext');
assert = require('assert');

/**
 * Retext.
 */

var retext;

retext = new Retext()
    .use(inspect)
    .use(link);

/**
 * Tests.
 */

describe('link()', function () {
    it('should be of type `function`', function () {
        assert(typeof link === 'function');
    });

    it('should classify `http://www.example.com` as a LinkNode',
        function (done) {
            retext.parse('Check out http://www.example.com!',
                function (err, tree) {
                    var node;

                    node = tree.head.head.tail.prev;

                    assert(node.type === 'LinkNode');
                    assert(node.toString() === 'http://www.example.com');

                    assert(node.data.protocol === 'http:');
                    assert(node.data.host === 'www.example.com');
                    assert(node.data.port === 80);
                    assert(node.data.hostname === 'www.example.com');
                    assert(node.data.hash === '');
                    assert(node.data.search === '');
                    assert(node.data.query === '');
                    assert(node.data.pathname === '/');
                    assert(node.data.path === '/');
                    assert(node.data.href === 'http://www.example.com/');

                    done(err);
                }
            );
        }
    );
});


spec/retext-link.spec.js
'use strict';

/**
 * Dependencies.
 */

var link,
    ast,
    visit,
    Retext,
    assert;

link = require('..');
ast = require('retext-ast');
visit = require('retext-visit');
Retext = require('retext');
assert = require('assert');

/**
 * Retext.
 */

var retext;

retext = new Retext()
    .use(visit)
    .use(ast)
    .use(link);

/**
 * Fixtures.
 */

var correctURLs,
    incorrectURLs;

correctURLs = [
    'http://foo.com/blah_blah',
    'http://foo.com/blah_blah/',
    'http://foo.com/blah_blah_(wikipedia)',
    'http://foo.com/blah_blah_(wikipedia)_(again)',
    'http://www.example.com/wpstyle/?p=364',
    'https://www.example.com/foo/?bar=baz&inga=42&quux',
    'http://✪df.ws/123',
    'http://userid:password@example.com:8080',
    'http://userid:password@example.com:8080/',
    'http://userid@example.com',
    'http://userid@example.com/',
    'http://userid@example.com:8080',
    'http://userid@example.com:8080/',
    'http://userid:password@example.com',
    'http://userid:password@example.com/',
    'http://142.42.1.1/',
    'http://142.42.1.1:8080/',
    'http://➡.ws/䨹',
    'http://⌘.ws',
    'http://⌘.ws/',
    'http://foo.com/blah_(wikipedia)#cite-1',
    'http://foo.com/blah_(wikipedia)_blah#cite-1',
    'http://foo.com/unicode_(✪)_in_parens',
    'http://foo.com/(something)?after=parens',
    'http://☺.damowmow.com/',
    'http://code.google.com/events/#&product=browser',
    'http://j.mp',
    'ftp://foo.bar/baz',
    'http://foo.bar/?q=Test%20URL-encoded%20stuff',
    'http://مثال.إختبار',
    'http://例子.测试',
    'http://उदाहरण.परीक्षा',
    'http://-.~_!$&\'()*+,;=:%40:80%2f::::::@example.com',
    'http://1337.net',
    'http://a.b-c.de',
    'http://223.255.255.254',

    /* Although not technically URL's, used a lot: */
    'foo.com/blah_blah',
    'foo.com/blah_blah/',
    'foo.com/blah_blah_(wikipedia)',
    'foo.com/blah_blah_(wikipedia)_(again)',
    'example.com/wpstyle/?p=364',
    'example.com/foo/?bar=baz&inga=42&quux',
    '✪df.ws/123',
    'userid:password@example.com:8080',
    'userid:password@example.com:8080/',
    'userid@example.com',
    'userid@example.com/',
    'userid@example.com:8080',
    'userid@example.com:8080/',
    'userid:password@example.com',
    'userid:password@example.com/',
    '142.42.1.1/',
    '142.42.1.1:8080/',
    '➡.ws/䨹',
    '⌘.ws',
    '⌘.ws/',
    'foo.com/blah_(wikipedia)#cite-1',
    'foo.com/blah_(wikipedia)_blah#cite-1',
    'foo.com/unicode_(✪)_in_parens',
    'foo.com/(something)?after=parens',
    '☺.damowmow.com/',
    'code.google.com/events/#&product=browser',
    'j.mp',
    'foo.bar/?q=Test%20URL-encoded%20stuff',
    'مثال.إختبار',
    '例子.测试',
    'उदाहरण.परीक्षा',
    '-.~_!$&\'()*+,;=:%40:80%2f::::::@example.com',
    '1337.net',
    'a.b-c.de',
    '223.255.255.254'
];

incorrectURLs = [
    '://',
    'http://',
    'http://.',
    'http://..',
    'http://../',
    'http://?',
    'http://??',
    'http://??/',
    'http://#',
    'http://##',
    'http://##/',
    '//',
    '//a',
    '///a',
    '///',
    'http:///a',
    'rdar://1234',
    'h://test',
    'http://3628126748'

    /* Although technically incorrect, I dont really care about these edge
     * cases:
     *
     * 'ftps://foo.bar/',
     * 'http://-error-.invalid/',
     * 'http://a.b--c.de/',
     * 'http://-a.b.co',
     * 'http://a.b-.co',
     * 'http://0.0.0.0',
     * 'http://10.1.1.0',
     * 'http://10.1.1.255',
     * 'http://224.1.1.1',
     * 'http://1.1.1.1.1',
     * 'http://123.123.123',
     * 'http://.www.foo.bar/',
     * 'http://www.foo.bar./',
     * 'http://.www.foo.bar./',
     * 'http://10.1.1.1',
     * 'http://10.1.1.254'
     */
];

/**
 * Tests.
 */

describe('link()', function () {
    it('should be of type `function`', function () {
        assert(typeof link === 'function');
    });

    it('should classify links (e.g., `www.example.com`) as a source node',
        function (done) {
            retext.parse('Check out www.example.com!', function (err, tree) {
                var node;

                node = tree.head.head.tail.prev;

                assert(node.toAST() === JSON.stringify({
                    'type' : 'SourceNode',
                    'value' : 'www.example.com',
                    'data' : {
                        'dataType' : 'link'
                    }
                }));

                done(err);
            });
        }
    );
});

/**
 * Validate a URL.
 *
 * @param {string} valueBefore
 * @param {string} valueAfter
 * @param {number} index - the location of the URL.
 * @param {string} url
 */

function validateURL(valueBefore, valueAfter, index, url) {
    it('should classify `' + url + '` as a source node', function (done) {
        retext.parse(valueBefore + url + valueAfter, function (err, tree) {
            var node;

            node = tree.head.head[index];

            assert(node.type === node.SOURCE_NODE);
            assert(node.data.dataType === 'link');
            assert(node.toString() === url);

            done(err);
        });
    });
}

/**
 * Validate a value is NOT classified as a URL.
 * Logs helpful messages.
 *
 * @param {string} valueBefore
 * @param {string} valueAfter
 * @param {number} index - the location of the non-URL.
 * @param {string} url
 */

function validateIncorrectURL(valueBefore, valueAfter, index, url) {
    it('should NOT classify `' + url + '` as a source node', function (done) {
        retext.parse(valueBefore + url + valueAfter, function (err, tree) {
            var node;

            node = tree.head.head[index];

            /**
             * Shouldn't be needed, but helpful for
             * future debugging.
             */

            /* istanbul ignore next */
            tree.visitType(tree.SOURCE_NODE, function (node) {
                console.log(
                    '    - A URL was, however, detected: ' + node
                );
            });

            assert(node.type !== node.SOURCE_NODE);
            assert(node.data.dataType !== 'link');

            done(err);
        });
    });
}

describe('Correct urls', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', ' it\'s awesome!', 4, correctURL);
    });
});

describe('Incorrect urls', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check out ', ' it\'s invalid', 4, incorrectURL);
    });
});

describe('Correct urls suffixed by a full-stop', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', '.', 4, correctURL);
    });
});

describe('Incorrect urls suffixed by a full-stop', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check the invalid URL ', '.', 8, incorrectURL);
    });
});

describe('Correct urls suffixed by a comma', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', ', or that.', 4, correctURL);
    });
});

describe('Incorrect urls suffixed by a comma', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check out ', ', its invalid.', 4, incorrectURL);
    });
});

describe('Correct urls suffixed by multiple full-stops', function () {
    correctURLs.forEach(function (correctURL) {
        validateURL('Check out ', '...', 4, correctURL);
    });
});

describe('Incorrect urls suffixed by multiple full-stops', function () {
    incorrectURLs.forEach(function (incorrectURL) {
        validateIncorrectURL('Check out ', '...', 4, incorrectURL);
    });
});


benchmark.js
'use strict';

var Retext,
    link;

/*
 * Dependencies.
 */

Retext = require('retext');
link = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithLink;

retext = new Retext();
retextWithLink = new Retext().use(link);

/*
 * Fixtures.
 *
 * Source:
 *   http://www.gutenberg.org/cache/epub/11024/pg11024.html
 */

var paragraph,
    linkParagraph;

/*
 * A paragraph, 5 sentences, filled with 10 links.
 */

linkParagraph = 'Thou art a churlish knight to so ' +
    'affront a http://example.com he could not sit ' +
    'upon http://userid:password@example.com/ his ' +
    'horse any longer . ' +

    'For methinks http://j.mp something hath befallen ' +
    'my lord and that he then, foo.com/blah_blah/ ' +
    'after a while, he cried out in great voice. ' +

    'For that 142.42.1.1:8080/ light in the sky lieth ' +
    'in the south a.b-c.de then Queen Helen fell down ' +
    'in a swoon, and lay. ' +

    'Touch me not, مثال.إختبار for I am not ' +
    'mortal, but Fay so the Lady of the Lake vanished ' +
    'away, उदाहरण.परीक्षा ' +
    'everything behind. ' +

    'Where she had http://⌘.ws stood was clear, and ' +
    'she was http://✪df.ws/123 gone since Sir Kay ' +
    'does not choose to assume my quarrel.';

/*
 * A paragraph, 5 sentences, without links.
 */

paragraph = 'Thou art a churlish knight to so affront ' +
    'a lady he could not sit upon his horse any ' +
    'longer. ' +

    'For methinks something hath befallen my lord ' +
    'and that he then, after a while, he cried out ' +
    'in great voice. ' +

    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and ' +
    'lay. ' +

    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, ' +
    'everything behind. ' +

    'Where she had stood was clear, and she was ' +
    'gone since Sir Kay does not choose to assume my ' +
    'quarrel.';

/*
 * Benchmark.
 */

suite('retext w/o retext-link', function () {
    bench('A paragraph (5 sentences, 100 words, 10 links)', function (done) {
        retext.parse(linkParagraph, done);
    });

    bench('A paragraph (5 sentences, 100 words, no links)', function (done) {
        retext.parse(paragraph, done);
    });
});

suite('retext w/ retext-link', function () {
    bench('A paragraph (5 sentences, 100 words, 10 links)', function (done) {
        retextWithLink.parse(linkParagraph, done);
    });

    bench('A paragraph (5 sentences, 100 words, no links)', function (done) {
        retextWithLink.parse(paragraph, done);
    });
});
